src/bootstrap/src/core/build_steps/run.rs

```diff
@@ -135,29 +133,35 @@ impl Step for Miri {
     }
 
     fn make_run(run: RunConfig<'_>) {
-        run.builder.ensure(Miri {
-            stage: run.builder.top_stage,
-            host: run.build_triple(),
-            target: run.target,
-        });
+        run.builder.ensure(Miri { target: run.target });
     }
 
     fn run(self, builder: &Builder<'_>) {
-        let stage = self.stage;
-        let host = self.host;
+        let host = builder.build.build;
         let target = self.target;
-        let compiler = builder.compiler(stage, host);
-
-        let miri =
-            builder.ensure(tool::Miri { compiler, target: self.host, extra_features: Vec::new() });
-        let miri_sysroot = test::Miri::build_miri_sysroot(builder, compiler, &miri, target);
+        let stage = builder.top_stage;
+        if stage == 0 {
+            eprintln!("miri cannot be run at stage 0");
+            std::process::exit(1);
+        }
+
+        // This compiler runs on the host, we'll just use it for the target.
+        let target_compiler = builder.compiler(stage, host);
+        // Similar to `compile::Assemble`, build with the previous stage's compiler. Otherwise
+        // we'd have stageN/bin/rustc and stageN/bin/rustdoc be effectively different stage
+        // compilers, which isn't what we want. Rustdoc should be linked in the same way as the
+        // rustc compiler it's paired with, so it must be built with the previous stage compiler.
+        let host_compiler = builder.compiler(stage - 1, host);
```

Comment:
@onur-ozkan I've changed the stage numbering here to match what rustdoc does: miri interprets the stage number as "the stage that this will be a compiler for", not "the stage that this is compiled with".

Conveniently this means `./x.py test miri` will finally avoid unnecessary work. Unfortunately this means `./x.py run miri` now uses the wrong default stage... but that's not used very often anyway.

Reply:
I'm not sure "faster, but wrong" is the right way to go...

Reply:
It's not faster but wrong though? It's exactly as correct and exactly as fast as before, the stage numbers just shifted by 1.

Previously I always did `./x.py test miri --stage 0`, now the `--stage 0` is no longer needed.

Reply:
Or do you mean the stage number is wrong now? According to which definition?

We have a dependency cycle here, and where we break the cycle is entirely arbitrary. All this change does is move the place where we break the cycle around a bit. Specifically Miri now follows rustc itself in that the 'test' step interprets the stage as "which stage are the programs that we are testing running in", which is always one higher than the stage that the compiler that builds/interprets them is built in.

Reply:
I'm referring to the behavior of `x run miri` using the wrong stage by default. That's what you said above right? `x test miri` now has the better default but `x run miri` has the wrong default; my concern is whether it's just slow or it's going to try to build Miri against the beta compiler.

Reply:
Ah, if it just produces a clear error message, that's probably par for bootstrap.

Reply:
> my concern is whether x run miri is just slow or is going to try to build Miri against the beta compiler.

Ah, sorry I misunderstood.
That will just print an error saying this doesn't work in stage 0.
Alternatively I could just make it pretend that you said stage 1...

Reply:
🤷 I don't have an opinion between those two options